schema.cc

```diff
@@ -437,6 +437,22 @@ schema::schema(const schema& o)
 {
 }
 
+schema::schema(reversed_tag, const schema& o)
+    : schema(o, [] (schema& s) {
+        // We invert the first 32bit octet of the MSB of the UUID, so doing two reversals gets us back to the original UUID
+        const int64_t mask = 0x00000000FFFFFFFF;
+        auto msb = s._raw._version.get_most_significant_bits();
+        msb = (~msb | mask) & (msb | ~mask);
```

Comment:
This will invert least significant bits of the timestamp. We would have a lower chance of collision if we inverted the most significant bits of the timestamp, since they change less often. Best to invert all the bits of the timestamp.

Reply:
Ok, will do.

Reply:
Collisions are prevented (probalistically) by the randomly generated part of the UUID, the chance of collision in this part is independent from how many bits we decide to flip, no?

More specifically: let X be an UUID, suppose the chance of obtaining X a second time is p. Then the chance of obtaining X with a single bit flipped (in the randomly generated part) is also p. (and so for any number of flipped bits)

So we could even flip a single bit (in the randomly generated part) and we'll be fine regarding collisions.
Do we need to flip more than one bit? Do we need to do it outside the randomly generated part for some reason?

Reply:
The random part is constant per shard (`const thread_local UUID_gen::clock_seq_and_node`). So lack of collision is guaranteed by two things, first that `clock_seq_and_node` is unique per shard globally, and second that timestamps increase monotonically, so are unique within `clock_seq_and_node`. Flipping lower bits of the timestamp may break uniqueness of the timestamps part within a given `clock_seq_and_node`.

Reply:
But that didn't really address your suggestion.

I think that flipping a bit in the random part would effectively create a new virtual node running on the same machine, on the same shard. More generally, we could use the same logic which is used by `make_clock_seq_and_node()`, which has additional checks, and generate a second set of ids for the purpose of generation of negated uuids. If make_clock_seq_and_node() generates a unique id, then negated uuids using a new clock_seq_and_node must be unique.

Reply:
We should hide this code in `UUID_gen.cc`, perhaps as `UUID_gen::negate(UUID)`